fs/readdircount_linux_amd64.go

```diff
@@ -0,0 +1,67 @@
+// +build linux,amd64
+
+package fs
+
+import (
+	"errors"
+	"os"
+	"syscall"
+	"unsafe"
+)
+
+func countDirEntries(buf []byte, n int) int {
+	count := 0
+	buf = buf[:n]
+	for len(buf) > 0 {
+		// see man page getdents(2) for struct linux_dirent64
+		reclen := *(*uint16)(unsafe.Pointer(&buf[16]))
+		ino := *(*uint64)(unsafe.Pointer(&buf[0]))
+
+		if int(reclen) > len(buf) {
+			return count
+		}
+		buf = buf[reclen:]
+		if ino == 0 {
+			continue
+		}
+		count++
+	}
+	return count
+}
+
+func getdents64(fd int, buf []byte) (n int, err int) {
+	var _p0 unsafe.Pointer
+	_p0 = unsafe.Pointer(&buf[0])
+	r0, _, errno := syscall.Syscall(syscall.SYS_GETDENTS64, uintptr(fd), uintptr(_p0), uintptr(len(buf)))
+	n = int(r0)
+	err = int(errno)
+	return
+}
+
+// ReadDirCount is similar to ReadDirNames() and then counting with len() but
+// it is optimized to avoid parsing the entries
+func (realFS) ReadDirCount(dir string) (int, error) {
+	buf := make([]byte, 4096)
```

Comment:
In a buffer of 4096 bytes, we can fit 170 entries (it's linear).

I tried the benchmarks with 2048 entries (increasing `ulimit -n` and opening `/dev/null` 2048 times to have that many entries in `/proc/self/fd`):
- len(buf)=64   ->  972097 ns/op
- len(buf)=256   ->  694169 ns/op
- len(buf)=512   ->  578525 ns/op
- len(buf)=1024 ->  570824 ns/op
- len(buf)=4096 ->  597399 ns/op
- len(buf)=8192 ->  573534 ns/op

Conclusion: that varies a bit but when the buffer is too small, the performance is negatively impacted. The buffer needs to be at least at 512 bytes (fitting ~21 directory entries), but increasing its size further does not improve the performance further much. I suggest to keep it at 4096.

I have not tried buffer pools yet.